### DISCUSSION TITLE
Hosting Supabase with a different (api.) subdomain and CloudFlare as DNS provider

### DISCUSSION THREAD

**Discussion by pyciko** [digitalocean/supabase-on-do#43] [Ideas]
It took me a day and a ton of help from @timothymamo to figure this out, so sharing with everyone else:)

**PART 1: Changing the subdomain**
You need to modify the subdomain in 5 places, namely

- In the docker-compose.yml file on [line 19](https://github.com/digitalocean/supabase-on-do/blob/main/packer/supabase/docker-compose.yml#L19)
- In the supabase.subdomain.conf file on [line 5](https://github.com/digitalocean/supabase-on-do/blob/main/packer/supabase/supabase.subdomain.conf#L5).
- In the .env.tfpl file on [line 62](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/files/.env.tftpl#L62) and [line 93](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/files/.env.tftpl#L93)
- In the network.tf file on [line 25](https://github.com/digitalocean/supabase-on-do/blob/main/terraform/network.tf#L25)

**PART 2: Using CloudFlare as DNS provider**
First of all, there's a small thing that I still haven't managed to fix. Even though we'll rely 100% on CloudFlare, the build process still asks DO API about the website. You need to register your "domain.com" in Networking -> Domains section and leave it there as it is, without any further actions. This is just to _suppress_ the error.

Create a new file in the [terraform/files](https://github.com/digitalocean/supabase-on-do/tree/main/terraform/files) directory called cloudflare.ini.tftpl with the following:
```
dns_cloudflare_email = ${TF_CLOUDFLARE_EMAIL}
dns_cloudflare_api_key = ${TF_CLOUDFLARE_API_KEY}
```

In the locals.tf file remove line 136 and add the following:
```
cloudflare_ini = templatefile("${path.module}/files/cloudflare.ini.tftpl",
    {
      TF_CLOUDFLARE_EMAIL = "${var.cloudflare_email}",
      TF_CLOUDFLARE_API_KEY = "${var.cloudflare_api_key}"
    }
  )
```
and change line 169 and 173 (164 and 168 before modification) in the locals.tf file to have the following:
```
path        = "/root/supabase/cloudflare.ini"
...
content     = base64encode("${local.cloudflare_ini}")

```
In your docker-compose.yml file change line 24 to the following:
`      - ./cloudflare.ini:/config/dns-conf/cloudflare.ini`

Now go to terraform/providers.tf file and, in required_providers, add 
```
cloudflare = {
      source  = "cloudflare/cloudflare"
      version = "~> 4.0"
    }
```
then, below, where providers are actually assigned with API keys, add this record:
```
provider "cloudflare" {
  email = var.cloudflare_email
  api_key = var.cloudflare_api_key
}
```
Now search for all occurrences of digitalocean_record (1 in network.tf and 5 in sendgrid.tf) and replace it with cloudflare_record, and also swap the domain parameter with "zone_id = var.cloudflare_zone_id" (this is a term used in CF).

Finally, go to variables.tf and add 3 new vars:
```
variable "cloudflare_email" {
  type        = string
}

variable "cloudflare_api_key" {
  type        = string
}

variable "cloudflare_zone_id" {
  type        = string
}

```
Zone ID is available in Dashboard, account API KEY is available in My Profile -> API TOKENS.
Now you're good to go, fill the variables in terraform.tfvars and deploy:)

**Comment by ndmikkelsen** [comment]
Also need to make sure to change line 20 of /packer/supabase/docker-compose.yml from "- DNSPLUGIN=digitalocean" to "- DNSPLUGIN=cloudflare"
